Implement 3-button/taiko navigation for main menu buttons

WebFreak001 opened the pull request:
I mainly intended to use this myself to navigate osu with a taiko controller, just starting out with the navigation screen here where it's quite simple to design.

For taiko controller usage it's intended to have left/right rim for left or right navigation and center drums as Select button. For a nicer experience on arrow keys, up and down acts pretty much the same as escape and enter respectively.

TODO: currently we reuse the hover effect for the focused main menu buttons. The styling should be made a bit more visible for keyboard navigation, since the hover isn't that visible afterwards if you didn't see the animation.

Sample video:

https://github.com/ppy/osu/assets/2035977/1f6ac201-41bd-4eee-8902-eb7306b9abf7

I think the hover effect / maybe only the keyboard focus effect could take design inspiration from previous iterations of lazer, like for example re-adding the gradient/glow around menu items like this:

**suggestion for the future** (please excuse the programmer art)

![image](https://github.com/ppy/osu/assets/2035977/c0d20594-4ce4-4874-8a58-39c5a29888d9)

WebFreak001 committed:
implement 3-button navigation for start of game

I mainly intended to use this myself to navigate osu with a taiko
controller, just starting out with the navigation screen here where it's
quite simple to design.

For taiko controller usage it's intended to have left/right rim for left
or right navigation and center drums as Select button. For a nicer
experience on arrow keys, up and down acts pretty much the same as
escape and enter respectively.

TODO: currently we reuse the hover effect for the focused main menu
buttons. The styling should be made a bit more visible for keyboard
navigation, since the hover isn't that visible afterwards if you didn't
see the animation.

WebFreak001 committed:
fix formatting

peppy commented on osu.Game.Tests/Visual/Navigation/TestSceneButtonSystemNavigation.cs:
```diff
@@ -16,6 +17,42 @@ public partial class TestSceneButtonSystemNavigation : OsuGameTestScene
 
+        private GlobalActionContainer globalActionContainer => Game.ChildrenOfType<GlobalActionContainer>().First();
+
+        [Test]
```
I'd definitely want to see test coverage of keyboard navigation getting interrupted by mouse, and then continuing from the last hovered button.

These are the kind of edge cases that require special attention for an implementation like this.

WebFreak001 committed:
fix CI issues

WebFreak001 committed:
fix build issues

WebFreak001 committed:
fix naming style violation

WebFreak001 committed:
fix indentation

WebFreak001 committed:
keyboard navigation fine tuning & mixed mouse test

WebFreak001 commented:
>  Error: Conditional access qualifier expression is known to be not null in osu.Game\Screens\Menu\ButtonSystem.cs on line 118
> Error: Expression is always null in osu.Game\Screens\Menu\ButtonSystem.cs on line 123
> Error: Conditional access qualifier expression is known to be null in osu.Game\Screens\Menu\ButtonSystem.cs on line 123

I don't think I can fix this, at least keeping the current code. The analyzer wrongly thinks that the property we access (`CurrentSelection`) is pure and doesn't change state. However after the `selectionIndex = value;`, the `CurrentSelection` property changes value, thus we can't have any assumptions that `previousButton` would always be true or false in the following cases looking at the new `CurrentSelection`.

WebFreak001 committed:
suppress possible null reference exception in test

WebFreak001 committed:
ignore hover lost when navigating with keyboard

bdach commented on osu.Game.Tests/Visual/Navigation/TestSceneButtonSystemNavigation.cs:
```diff
@@ -16,6 +18,103 @@ public partial class TestSceneButtonSystemNavigation : OsuGameTestScene
+        [Test]
+        public void TestKeyboardNavigation()
+        {
+            AddAssert("state is initial", () => buttons.State == ButtonSystemState.Initial);
```
We prefer to write assertions in nunit matcher style these days.

```suggestion
            AddAssert("state is initial", () => buttons.State, () => Is.EqualTo(ButtonSystemState.Initial));
```

Brings better error messaging and all that. I'm not commenting this everywhere, hopefully you get the idea.

bdach commented on osu.Game/Screens/Menu/ButtonSystem.cs:
```diff
@@ -81,10 +81,66 @@ public void SetOsuLogo(OsuLogo logo)
         private readonly MainMenuButton backButton;
+        private readonly MainMenuButton settingsButton;
+
+        // public for tests
+        [CanBeNull]
+        public MainMenuButton PlayButton { get; private set; }
```
Doesn't need to be exposed for testing - can be exposed with `.ChildrenOfType<MainMenuButton>().First()` or similar.

WebFreak001 replied:
I decided to keep playButton, but make it internal, since it makes the usage in ButtonSystem a bit more clearer, as well as the tests. Additionally it's not the first child, it's somewhere in the middle, since the back and settings button come before it.

bdach commented on osu.Game/Screens/Menu/ButtonSystem.cs:
```diff
@@ -81,10 +81,66 @@ public void SetOsuLogo(OsuLogo logo)
         private readonly List<MainMenuButton> buttonsTopLevel = new List<MainMenuButton>();
         private readonly List<MainMenuButton> buttonsPlay = new List<MainMenuButton>();
 
+        [CanBeNull]
+        public List<MainMenuButton> CurrentButtonsList => State switch
+        {
+            ButtonSystemState.TopLevel => buttonsTopLevel,
+            ButtonSystemState.Play => buttonsPlay,
+            _ => null
+        };
```
We prefer plain old `switch` blocks to switchexprs around these parts.

bdach commented on osu.Game/Screens/Menu/ButtonSystem.cs:
```diff
@@ -81,10 +81,66 @@ public void SetOsuLogo(OsuLogo logo)
+            get => selectionIndex;
+            set
+            {
+                MainMenuButton previousButton = null;
+                if (CurrentSelection is MainMenuButton button)
+                    previousButton = button;
+
+                selectionIndex = value;
+
+                if (CurrentSelection is MainMenuButton newButton)
+                {
+                    if (previousButton != newButton)
+                    {
+                        previousButton?.SimulateHoverLost();
+                        newButton.SimulateHover();
+                    }
+                }
+                else
+                    previousButton?.SimulateHoverLost();
```
All this is a bit horrific but I don't see a great way of replacing it. That said, you could do

```suggestion
                var previousButton = CurrentSelection as MainMenuButton;
                previousButton?.SimulateHoverLost();

                selectionIndex = value;

                var currentButton = CurrentSelection as MainMenuButton;
                currentButton?.SimulateHover();
```

which doesn't seem to confuse static analysis, I guess.

WebFreak001 replied:
yeah I wrote it this way, so it doesn't call SimulateHoverLost and SimulateHover on the same object, to avoid it confusing the animations. I think this can mostly be skipped by checking for `selectionIndex != value` though

bdach commented on osu.Game/Screens/Menu/ButtonSystem.cs:
```diff
@@ -81,10 +81,66 @@ public void SetOsuLogo(OsuLogo logo)
+            _ => null
+        };
+
+        private int selectionIndex = -1;
```
Would really like this to be `int?` and for `null` to denote the "no selection" case rather than use C-style special values.

bdach commented on osu.Game/Screens/Menu/ButtonSystem.cs:
```diff
@@ -81,10 +81,66 @@ public void SetOsuLogo(OsuLogo logo)
+        /// input. Reset to Vector2.Zero when hovering over buttons. When a
+        /// hover is registered very close to this position, we ignore the hover.
+        /// </summary>
+        private Vector2 blockedMousePosition = Vector2.Zero;
```
Please use `Vector2?` and `null` as the special "hovering over buttons" value instead. Inline "guard" values like this `Vector2.Zero` that could be actual valid values are evil.

bdach commented on osu.Game/Screens/Menu/ButtonSystem.cs:
```diff
@@ -241,10 +303,31 @@ public bool OnPressed(KeyBindingPressEvent<GlobalAction> e)
+                    SelectUp();
+                    return true;
+
+                case GlobalAction.SelectNextGroup:
```
Really not sure about hooking this keybinding up to these actions. `Select{Previous,Next}` are okay, but there is no grouping to be had here in sight.

WebFreak001 replied:
you are right, this was actually more of a leftover from my previous implementation where (down and right) and (up and left) were pairs of the same action. I think for simple keyboard navigation the current approach is good, but actually broke my taiko use-case which I initially made this for without even noticing.

Having a new set of definable buttons (FocusNext, FocusPrevious, EnterFocus) for this navigation would make it much more understandable as well as opening up the opportunity to implement this in other views as well.

WebFreak001 replied:
ok so after removing SelectPrevious and SelectNext, suddenly `Select{Previous,Next}Group` made much more sense in the code again and I decided to not add any new bindings here, and instead promote reusing next/previous group for focus group change. Will push in a sec

bdach commented on osu.Game/Screens/Menu/ButtonSystem.cs:
```diff
@@ -256,6 +339,96 @@ public void OnReleased(KeyBindingReleaseEvent<GlobalAction> e)
+            SelectionIndex = parentSelection;
+        }
+
+        public void ClearKeyboardSelection(HoverEvent e)
```
Would much prefer written as

```csharp
        public void ClearKeyboardSelection(HoverEvent e)
        {
            SelectionIndex = -1;
            if (CurrentButtonsList == null)
                return;

            foreach (var btn in CurrentButtonsList)
            {
                if (btn != e.Target)
                {
                    btn.SimulateHoverLost();
                }
            }
        }
```

Similar applies to `PreferKeyboardNavigation()`.

bdach commented on osu.Game/Screens/Menu/MainMenuButton.cs:
```diff
@@ -48,15 +49,22 @@ public partial class MainMenuButton : BeatSyncedContainer, IStateful<ButtonState
         private Sample sampleClick;
         private Sample sampleHover;
 
+        // for tests
+        public bool LooksHovered { get; private set; }
```
This should not really need to exist. But if it has to, at least as `internal`.

bdach commented on osu.Game/Screens/Menu/MainMenuButton.cs:
```diff
@@ -48,15 +49,22 @@ public partial class MainMenuButton : BeatSyncedContainer, IStateful<ButtonState
         public override bool ReceivePositionalInputAt(Vector2 screenSpacePos) => box.ReceivePositionalInputAt(screenSpacePos);
 
-        public MainMenuButton(LocalisableString text, string sampleName, IconUsage symbol, Color4 colour, Action clickAction = null, float extraWidth = 0, Key triggerKey = Key.Unknown)
+        public MainMenuButton(ButtonSystem buttonSystem, LocalisableString text, string sampleName, IconUsage symbol, Color4 colour, Action clickAction = null, float extraWidth = 0,
```
I really don't like that the buttons receive the entire `ButtonSystem` and call check methods on it. The `ConfirmHover()` thingy can be done by having each `MainMenuButton` expose a `HandleHover` bindable or something, and binding it to a single source of truth that would sit in `ButtonSystem`.

WebFreak001 replied:
since i introduced the hover action anyway, I made it now trigger for hover and unhover with position and target, which I think cleaned this up a lot and also made ButtonSystem nicer to read. Thanks for the suggestions.

WebFreak001 committed:
adjust to review

WebFreak001 committed:
fix naming violations

WebFreak001 commented:
will squash once reviews are done

bdach commented:
Please don't squash. Force pushing is also not usually done, while we're at it. "Clean history" is not something we hold in high regard here.

WebFreak001 committed:
fix nvika test results

WebFreak001 commented:
> Please don't squash. Force pushing is also not usually done, while we're at it. "Clean history" is not something we hold in high regard here.

since I push quite frequently to run the CI tests, I would at least want to squash together the simple "fix formatting", "fix CI" commits onto the last real commit, just because it's a personal preference.

The CI pipelines / the linter script doesn't seem to run on my machine in a reasonable amount of time, so CI is much faster for me.

WebFreak001 committed:
fix nvika formatting issues again

WebFreak001 committed:
remove up/down navigation from the menu

it only really worked as expected with arrow keys and would immediately
feel weird with custom keybindings.

To make this feel more consistent with existing focus changes and future
focus changes, I removed these entirely.

peppy commented:
> The CI pipelines / the linter script doesn't seem to run on my machine in a reasonable amount of time, so CI is much faster for me.

That may be, but the builds delay all other actions on the ppy organisation if you do it after PRing.

If you're going to rely on github for CI, I'd at least recommend doing it before PRing in the future as then it will allocate to your own account.

WebFreak001 commented:
> That may be, but the builds delay all other actions on the ppy organisation if you do it after PRing.
>
> If you're going to rely on github for CI, I'd at least recommend doing it before PRing in the future as then it will allocate to your own account.

sorry, I forgot about that and thought the actions here would run on my account's quota. (since in github actions secrets are also coming from the fork user's repository)

I'm not quite sure what was going on with my local setup though, since CPU usage was just maxed out for like 15 minutes, then I just cancelled the linting attempt without any output. Some tools didn't run, since it was complaining about a missing runtime, which might have been dotnet core 3 that I don't have installed, but am seeing in the CI scripts. Might work better on my better spec'd machine than on my laptop though.

bdach commented:
Test failures need looking into.

peppy commented:
Ignoring the code for a moment (it has some very serious legibility issues which may block it), the behaviour doesn't match expectations.

Take this example where I'm intercepting keyboard selection with the mouse:

https://github.com/ppy/osu/assets/191335/b50bcc3c-29b9-477f-8184-f937eef62004

You'd expect the selection to change from the actual hovered selection – not a separate invisible selection index.

Other issues:

- Repeats are not handled, but should be
- Hover state doesn't update after entering keyboard selection until unhovering / rehovering:


https://github.com/ppy/osu/assets/191335/b3576d0a-b8cf-4f54-a4df-4295b1e4460d

- I'm not sure the selection resetting on returning from an overlay makes sense:

https://github.com/ppy/osu/assets/191335/fbc65eb4-e4cb-4cc6-8737-bfcfb55a391f

All this would need test coverage. Also tests are failing.

I wasn't going to put too much thought into code quality when reviewing this because the whole menu is intended to be replaced in the near future, but I think these issues would need to be fixed at very least.

peppy commented on osu.Game/Screens/Menu/ButtonSystem.cs:
```diff
@@ -286,7 +460,7 @@ private bool onOsuLogo()
 
                 case ButtonSystemState.TopLevel:
-                    buttonsTopLevel.First().TriggerClick();
+                    PlayButton!.TriggerClick();
```
Please don't use `!`. Especially after marking the field `CanBeNull`. Check other files for examples of how this should be done, and remove the `nullable disable` rather than adding new jetbrains annotations.

peppy commented on osu.Game/Screens/Menu/MainMenuButton.cs:
```diff
@@ -158,22 +164,41 @@ protected override void OnNewBeat(int beatIndex, TimingControlPoint timingPoint,
+            return true;
+        }
+
+        public void SimulateHover()
```
What does this even mean?

peppy commented on osu.Game/Screens/Menu/ButtonSystem.cs:
```diff
@@ -127,23 +184,36 @@ public ButtonSystem()
             buttonArea.AddRange(buttonsTopLevel);
 
+            // for keyboard navigation
+            buttonsTopLevel.Insert(0, settingsButton);
+            buttonsPlay.Insert(0, backButton);
```
The comment you added here doesn't do anything to explain what this is actually doing.

peppy reviewed (commented):
Basic review
